src/API/Google/BudgetMetrics.php

```diff
@@ -0,0 +1,145 @@
+use Google\Ads\GoogleAds\V18\Services\GenerateRecommendationsRequest\BiddingInfo;
+use Google\Ads\GoogleAds\V18\Services\GenerateRecommendationsRequest\BudgetInfo;
+use Google\ApiCore\ApiException;
+
+/**
+ * Class BudgetMetrics
+ *
+ * @package Automattic\WooCommerce\GoogleListingsAndAds\API\Google
+ */
+class BudgetMetrics implements OptionsAwareInterface, TransientsAwareInterface {
+
+	use MicroTrait;
+	use OptionsAwareTrait;
+	use PluginHelper;
+	use TransientsAwareTrait;
+	use LocationIDTrait;
+
+	/**
+	 * The Google Ads Client.
+	 *
+	 * @var GoogleAdsClient
+	 */
+	protected $client;
+
+	/**
+	 * BudgetMetrics constructor.
+	 *
+	 * @param GoogleAdsClient $client
+	 */
+	public function __construct( GoogleAdsClient $client ) {
+		$this->client = $client;
+	}
+
+	/**
+	 * Fetch budget metrics from Google Ads API.
+	 *
+	 * @param float $budget        Budget to fetch metrics for.
+	 * @param array $country_codes List of countries to include.
+	 *
+	 * @return array|null List of metrics.
+	 */
+	public function get_metrics( float $budget, array $country_codes ): ?array {
+		$cache_key = strtolower( join( '-', $country_codes ) . '-' . $budget );
+		$transient = $this->transients->get( TransientsInterface::ADS_BUDGET_METRICS );
+
+		// Check if we have the budget metrics cached in the transient.
+		if ( $transient && ! empty( $transient[ $cache_key ] ) ) {
+			return $transient[ $cache_key ];
+		}
+
+		$location_ids = $this->get_location_ids( $country_codes );
+
+		$request = new GenerateRecommendationsRequest(
+			[
+				'customer_id'              => $this->options->get_ads_id(),
+				'recommendation_types'     => [ RecommendationType::CAMPAIGN_BUDGET ],
+				'advertising_channel_type' => AdvertisingChannelType::PERFORMANCE_MAX,
+				'positive_locations_ids'   => array_keys( $location_ids ),
+				'country_codes'            => $country_codes,
+				'bidding_info'             => new BiddingInfo(
+					[
+						'bidding_strategy_type' => BiddingStrategyType::MAXIMIZE_CONVERSION_VALUE,
+					]
+				),
+				'budget_info'              => new BudgetInfo(
+					[
+						'current_budget' => $this->to_micro( $budget ),
+					],
+				),
+				'asset_group_info'         => [
+					new AssetGroupInfo(
+						[
+							'final_url' => $this->get_site_url(),
+						],
+					),
+				],
+			]
+		);
+
+		try {
+			$response = $this->client->getRecommendationServiceClient()->generateRecommendations( $request );
+
+			foreach ( $response->getRecommendations() as $recommendation ) {
+				$campaign_budget_recommendation = $recommendation->getCampaignBudgetRecommendation();
+				if ( ! $campaign_budget_recommendation ) {
+					continue;
+				}
+
+				// Parse the metrics for the given budget.
+				$metrics = $this->parse_metrics( $campaign_budget_recommendation, $budget );
+				$this->transients->set( TransientsInterface::ADS_BUDGET_METRICS, [ $cache_key => $metrics ], HOUR_IN_SECONDS * 12 );
```

Comment:
This cache always stores only the last query result that arrived here. I wonder is this the intended implementation?

Consider that the user may adjust the value back and forth when editing the budget. Although the frontend can add a debounce to reduce the number of calls to the API, the cache won't be hit even if the user is just switching between two values.

Reply:
Good suggestion. We pulled the previous value of the transient already, so the two should be merged, to cache multiple values with unique cache keys.